(#3539) Add design documentation for C# cmdlets

vexx32 opened the pull request:
## Description Of Changes

- Added CONTRIBUTING.md document to the Chocolatey.PowerShell project.

## Motivation and Context

Given we use some slightly custom bits in Chocolatey.PowerShell that will alter how contributors are expected to work with the cmdlets in this project compared to how more bare-bones projects handle C# cmdlets, this document outlines some of the more notable changes as well as the common design practices we will be using here.

## Testing

N/A, doc only.

### Operating Systems Testing

N/A

## Change Types Made

* [ ] Bug fix (non-breaking change).
* [ ] Feature / Enhancement (non-breaking change).
* [ ] Breaking change (fix or feature that could cause existing functionality to change).
* [x] Documentation changes.
* [ ] PowerShell code changes.

## Change Checklist

* [ ] Requires a change to the documentation.
* [ ] Documentation has been updated.
* [ ] Tests to cover my changes, have been added.
* [ ] All new and existing tests passed?
* [ ] PowerShell code changes: PowerShell v3 compatibility checked?

## Related Issue
<!-- Make sure you have raised an issue for this pull request before
continuing. -->

Fixes #3539

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,100 @@
+
+### Output
+
+By default, `ChocolateyCmdlet`'s `WriteObject()` method will enumerate collections when outputting them, similar to how PowerShell's `Write-Output` works by default. If you need to disable this, use the `WriteObject `
```
This line seems to end abruptly... I suspect there's more to be said here.

vexx32 replied:
Ah, good catch, thanks. Must have meant to look up what the overload is and didn't come back to it.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,100 @@
+
+### Logging
+
+Cmdlets inheriting from `ChocolateyCmdlet` will log their parameter values to debug logs when called by default. If any parameters may contain sensitive information, override the `Logging` property and set it to `false` to disable this behaviour.
```
Is there any logging that should be done in these cases? Will this be taken care of automatically?

vexx32 replied:
At _present_ I would say no. Currently the commands that work this way do not get logged at all I don't think, from the ones I've seen so far?

We could safely turn on the logging after a cmdlet runs (which I think for the commands that have it only mentions the command _name_ and nothing else) and have this property only toggle the logging of parameter values at the start, but I think that would be another divergence in behaviour and we should evaluate that after we finish the rewrite, given I think it has pretty minimal impact whether we log a command after it runs or not.

Most of the commands that do this are... not really part of the public API anyway.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,100 @@
+
+By default, `ChocolateyCmdlet`'s `WriteObject()` method will enumerate collections when outputting them, similar to how PowerShell's `Write-Output` works by default. If you need to disable this, use the `WriteObject `
+
+### Helpers
+
+`ChocolateyCmdlet` provides some helper methods for common operations that might be needed for many cmdlets.
+Some of these (and many more) are available on the `PSHelper` class.
```
Is there more to this section we want to include in this section? Is the intention that the next section answers the questions about what kind of things might be here? Should that perhaps be mentioned?

vexx32 replied:
Mmmm. I think most of that is answered in the next section.... the exact line for what goes onto `ChocolateyCmdlet` _itself_ is a little fuzzier and I don't have a fantastic way to delineate that really... In theory anything that takes `PSCmdlet` as a parameter _could_ be a helper method attached to `ChocolateyCmdlet`, but in practice I think the useful set of things to have directly available is a good bit narrower given calling into `PSHelper` is not really a big ask.

corbob reviewed (changes requested):
Overall I think this looks good, just have a few clarifying questions, at least one I think requires changes.

corbob commented:
One other thing that's occurred to me: Should we consider linking to this from the higher up contributing document? I'm a little concerned about the discoverability of the document if it's just a file in a directory where you might not look for it as you're in Visual Studio...

vexx32 commented:
> Should we consider linking to this from the higher up contributing document?

That's something I thought about and... am not really sure about. I suppose it can't hurt, I did have a bit of a difficult time trying to figure out a sensible place in the main CONTRIBUTING doc to reference it, but I can always add a new section...

It also should probably be put into the sln file just to make it more visible in VS, I think. I'll do both.

gep13 reviewed (approved):
LGTM!

gep13 commented:
@vexx32 just had a read through this, and this reads really well to me, and it covers the things that we have spoken about, so I would be happy to see this merged in.

I will leave @corbob to come back and confirm based on any changes that he asked for.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,101 @@
+
+### Output
+
+By default, `ChocolateyCmdlet`'s `WriteObject()` method will enumerate collections when outputting them, similar to how PowerShell's `Write-Output` works by default.
```
```suggestion
By default, `ChocolateyCmdlet`'s `WriteObject(obj)` method will enumerate collections when outputting them, similar to how PowerShell's `Write-Output` works by default.
```

It's probably a good idea to specify using this overload as opposed to one with no parameters.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,101 @@
+### Output
+
+By default, `ChocolateyCmdlet`'s `WriteObject()` method will enumerate collections when outputting them, similar to how PowerShell's `Write-Output` works by default.
+If you need to disable this, use the `WriteObject(obj, false)` overload.
```
I presume that this overload is not yet merged into the target branch, but will be before it's merged into `develop`. Part of me wonders if the `false` should be identified as to what it's parameter is as just reading this document it's perhaps a little unclear.

vexx32 replied:
No, that overload is a default one inherited from `Cmdlet` or `PSCmdlet`, not one we're adding. I can add the parameter name.

corbob reviewed (changes requested):
Just a few minor suggestions.

vexx32 commented:
@corbob I have incorporated your suggestions, and added a section about helpers to indicate they should not be directly writing to PowerShell output/error streams and suggest a common pattern to be used for those cases.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,161 @@
+> To facilitate this, we can put the core logic of `Install-ChocolateyInstallPackage` into a helper class, and the actual `InstallChocolateyInstallPackageCommand` class we would write would only define parameters, then call into the helper class.
+> Then, when we write the `InstallChocolateyPackageCommand`, it can download files and then call into the same helper class to run the other cmdlet's logic seamlessly.
+
+### Helpers should return data or throw exceptions, not write output or errors
```
For completeness @gep13 could you review this section as it was added after your approval.

corbob commented on src/Chocolatey.PowerShell/CONTRIBUTING.md:
```diff
@@ -0,0 +1,161 @@
+Additionally, since all cmdlets are initialised by the PowerShell runtime, we cannot use a dependency injection framework to make things work.
+As a result, many of the helpers will be a little more bare-bones and not make as heavy use of interfaces as other code areas in this repository.
+
+## Interop with Chocolatey CLI
```
For completeness @gep13 could you review this section as it was added after your approval.

corbob reviewed (approved):
These changes look good to me. I've tagged @gep13 on the headings of the sections that are new since his review.

gep13 reviewed (approved):
LGTM!

corbob commented:
The GHA build is due to #3607. Merging this in as it's targetting a feature branch and not develop.

corbob commented:
Thanks for getting this fixed up @vexx32 👍
